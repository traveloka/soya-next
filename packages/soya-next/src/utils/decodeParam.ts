export default function decodeParam<TParam>(
  param: TParam
): TParam extends string ? string : TParam;
export default function decodeParam<TParam>(param: TParam) {
  if (typeof param !== "string" || param.length === 0) {
    return param;
  }

  try {
    return decodeURIComponent(param);
  } catch (err) {
    // istanbul ignore else
    if (err instanceof URIError) {
      err.message = `Failed to decode param "${param}"`;
      err.status = err.statusCode = 400;
    }
    throw err;
  }
}
