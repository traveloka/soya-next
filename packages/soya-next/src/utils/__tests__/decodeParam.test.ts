import decodeParam from "../decodeParam";

describe("DecodeParam Utils", () => {
  it("should return as is if it's not a string or is an empty string", () => {
    expect(decodeParam("")).toBe("");
    expect(decodeParam({ a: 1 })).toMatchSnapshot();
  });

  it("should throw an URIError exception when used wrongly", () => {
    expect(() => {
      decodeParam("/error/%A0");
    }).toThrowErrorMatchingSnapshot();
  });
});
