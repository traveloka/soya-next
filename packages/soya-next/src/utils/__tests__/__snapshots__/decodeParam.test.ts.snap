// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DecodeParam Utils should return as is if it's not a string or is an empty string 1`] = `
Object {
  "a": 1,
}
`;

exports[`DecodeParam Utils should throw an URIError exception when used wrongly 1`] = `"Failed to decode param \\"/error/%A0\\""`;
